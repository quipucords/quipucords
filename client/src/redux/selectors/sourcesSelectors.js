import { createSelector } from 'reselect';
import _get from 'lodash/get';
import helpers from '../../common/helpers';
import apiTypes from '../../constants/apiConstants';

/**
 * Map a new source object to consumable prop names
 */
const sourceDetail = state => state.addSourceWizard.source;

/**
 * Map an edit source object to consumable prop names
 */
const editSourceDetail = state => state.addSourceWizard.editSource;

const sourceDetailSelector = createSelector(
  [sourceDetail, editSourceDetail],
  (source, editSource) => {
    const updateSource = Object.assign({}, editSource || {}, source || {});
    const newSource = {};

    if (updateSource) {
      /**
       * Allow for an edit source with credentials in the form of
       * [{ id:Number, type:String, name:String }]
       * Or a new source in the form of
       * [id]
       */
      helpers.setPropIfDefined(
        newSource,
        ['credentials'],
        _get(updateSource, apiTypes.API_RESPONSE_SOURCE_CREDENTIALS, []).map(
          cred => cred[apiTypes.API_RESPONSE_SOURCE_CREDENTIALS_ID] || cred
        )
      );

      helpers.setPropIfDefined(newSource, ['hosts'], updateSource[apiTypes.API_RESPONSE_SOURCE_HOSTS]);
      helpers.setPropIfDefined(newSource, ['id'], updateSource[apiTypes.API_RESPONSE_SOURCE_ID]);
      helpers.setPropIfDefined(newSource, ['name'], updateSource[apiTypes.API_RESPONSE_SOURCE_NAME]);

      if (updateSource[apiTypes.API_RESPONSE_SOURCE_OPTIONS]) {
        helpers.setPropIfDefined(
          newSource,
          ['optionSslCert'],
          _get(updateSource[apiTypes.API_RESPONSE_SOURCE_OPTIONS], apiTypes.API_RESPONSE_SOURCE_OPTIONS_SSL_CERT)
        );

        helpers.setPropIfDefined(
          newSource,
          ['optionSslProtocol'],
          _get(updateSource[apiTypes.API_RESPONSE_SOURCE_OPTIONS], apiTypes.API_RESPONSE_SOURCE_OPTIONS_SSL_PROTOCOL)
        );

        helpers.setPropIfDefined(
          newSource,
          ['optionDisableSsl'],
          _get(updateSource[apiTypes.API_RESPONSE_SOURCE_OPTIONS], apiTypes.API_RESPONSE_SOURCE_OPTIONS_DISABLE_SSL)
        );

        helpers.setPropIfDefined(
          newSource,
          ['optionParamiko'],
          _get(updateSource[apiTypes.API_RESPONSE_SOURCE_OPTIONS], apiTypes.API_RESPONSE_SOURCE_OPTIONS_PARAMIKO)
        );
      }

      helpers.setPropIfDefined(newSource, 'port', updateSource[apiTypes.API_RESPONSE_SOURCE_PORT]);
      helpers.setPropIfDefined(newSource, ['type'], updateSource[apiTypes.API_RESPONSE_SOURCE_SOURCE_TYPE]);

      if (newSource.hosts && newSource.hosts.length) {
        newSource.hostsMultiple = newSource.hosts.join(',\n');
        newSource.hostsSingle = `${newSource.hosts[0]}:${newSource.port || ''}`;
      }
    }

    return newSource;
  }
);

const makeSourceDetailSelector = () => sourceDetailSelector;

const sourcesSelectors = {
  sourceDetail: sourceDetailSelector,
  makeSourceDetail: makeSourceDetailSelector
};

export { sourcesSelectors as default, sourcesSelectors, sourceDetailSelector, makeSourceDetailSelector };
