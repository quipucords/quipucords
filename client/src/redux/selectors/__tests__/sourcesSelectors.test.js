import sourcesSelectors from '../sourcesSelectors';
import apiTypes from '../../../constants/apiConstants';

describe('SourcesSelectors', () => {
  it('Should return specific selectors', () => {
    expect(sourcesSelectors).toMatchSnapshot('selectors');
  });

  it('Should map a source to consumable props', () => {
    const state = {
      addSourceWizard: {
        source: {
          [apiTypes.API_RESPONSE_SOURCE_CREDENTIALS]: [15],
          [apiTypes.API_RESPONSE_SOURCE_HOSTS]: ['192.168.0.1'],
          [apiTypes.API_RESPONSE_SOURCE_ID]: 1,
          [apiTypes.API_RESPONSE_SOURCE_NAME]: 'lorem',
          [apiTypes.API_RESPONSE_SOURCE_OPTIONS]: {
            [apiTypes.API_RESPONSE_SOURCE_OPTIONS_PARAMIKO]: false,
            [apiTypes.API_RESPONSE_SOURCE_OPTIONS_DISABLE_SSL]: false,
            [apiTypes.API_RESPONSE_SOURCE_OPTIONS_SSL_PROTOCOL]: false,
            [apiTypes.API_RESPONSE_SOURCE_OPTIONS_SSL_CERT]: false
          },
          [apiTypes.API_RESPONSE_SOURCE_PORT]: 22,
          [apiTypes.API_RESPONSE_SOURCE_SOURCE_TYPE]: 'network'
        }
      }
    };

    expect(sourcesSelectors.sourceDetail(state)).toMatchSnapshot('sourceDetail');

    const updatedState = {
      addSourceWizard: {
        source: {
          [apiTypes.API_RESPONSE_SOURCE_CREDENTIALS]: [
            {
              [apiTypes.API_RESPONSE_SOURCE_CREDENTIALS_ID]: 10
            }
          ],
          [apiTypes.API_RESPONSE_SOURCE_HOSTS]: ['192.168.0.1'],
          [apiTypes.API_RESPONSE_SOURCE_ID]: 1,
          [apiTypes.API_RESPONSE_SOURCE_NAME]: 'lorem',
          [apiTypes.API_RESPONSE_SOURCE_OPTIONS]: {
            [apiTypes.API_RESPONSE_SOURCE_OPTIONS_PARAMIKO]: false,
            [apiTypes.API_RESPONSE_SOURCE_OPTIONS_DISABLE_SSL]: false,
            [apiTypes.API_RESPONSE_SOURCE_OPTIONS_SSL_PROTOCOL]: false,
            [apiTypes.API_RESPONSE_SOURCE_OPTIONS_SSL_CERT]: false
          },
          [apiTypes.API_RESPONSE_SOURCE_PORT]: 22,
          [apiTypes.API_RESPONSE_SOURCE_SOURCE_TYPE]: 'network'
        }
      }
    };

    expect(sourcesSelectors.sourceDetail(updatedState)).toMatchSnapshot('sourceDetail updated');
  });

  it('Should minimally map an undefined source or properties', () => {
    const state = {
      addSourceWizard: {
        source: undefined
      }
    };

    expect(sourcesSelectors.sourceDetail(state)).toMatchSnapshot('sourceDetail undefined');

    state.addSourceWizard.source = {
      [apiTypes.API_RESPONSE_SOURCE_CREDENTIALS]: undefined,
      [apiTypes.API_RESPONSE_SOURCE_HOSTS]: undefined,
      [apiTypes.API_RESPONSE_SOURCE_ID]: undefined,
      [apiTypes.API_RESPONSE_SOURCE_NAME]: undefined,
      [apiTypes.API_RESPONSE_SOURCE_OPTIONS]: {
        [apiTypes.API_RESPONSE_SOURCE_OPTIONS_PARAMIKO]: undefined,
        [apiTypes.API_RESPONSE_SOURCE_OPTIONS_DISABLE_SSL]: undefined,
        [apiTypes.API_RESPONSE_SOURCE_OPTIONS_SSL_PROTOCOL]: undefined,
        [apiTypes.API_RESPONSE_SOURCE_OPTIONS_SSL_CERT]: undefined
      },
      [apiTypes.API_RESPONSE_SOURCE_PORT]: undefined,
      [apiTypes.API_RESPONSE_SOURCE_SOURCE_TYPE]: undefined
    };

    expect(sourcesSelectors.sourceDetail(state)).toMatchSnapshot('sourceDetail properties undefined');
  });
});
