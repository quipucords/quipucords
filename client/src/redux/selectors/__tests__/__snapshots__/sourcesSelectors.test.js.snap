// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`selectors 1`] = `
Object {
  "makeSourceDetail": [Function],
  "sourceDetail": [Function],
}
`;

exports[`sourceDetail 1`] = `
Object {
  "credentials": Array [
    15,
  ],
  "hosts": Array [
    "192.168.0.1",
  ],
  "hostsMultiple": "192.168.0.1",
  "hostsSingle": "192.168.0.1:22",
  "id": 1,
  "name": "lorem",
  "optionDisableSsl": false,
  "optionParamiko": false,
  "optionSslCert": false,
  "optionSslProtocol": false,
  "port": 22,
  "type": "network",
}
`;

exports[`sourceDetail properties undefined 1`] = `
Object {
  "credentials": Array [],
}
`;

exports[`sourceDetail undefined 1`] = `
Object {
  "credentials": Array [],
}
`;

exports[`sourceDetail updated 1`] = `
Object {
  "credentials": Array [
    10,
  ],
  "hosts": Array [
    "192.168.0.1",
  ],
  "hostsMultiple": "192.168.0.1",
  "hostsSingle": "192.168.0.1:22",
  "id": 1,
  "name": "lorem",
  "optionDisableSsl": false,
  "optionParamiko": false,
  "optionSslCert": false,
  "optionSslProtocol": false,
  "port": 22,
  "type": "network",
}
`;
